cms/djangoapps/models/settings/course_metadata.py

```diff
@@ -282,6 +286,57 @@ def validate_and_update_from_json(cls, block, jsondict, user, filter_tabs=True):
 
         return did_validate, errors, updated_data
 
+    @staticmethod
+    def get_user_partition_id(block, min_partition_id, max_partition_id):
+        """
+        Get a dynamic partition id that is not already in use.
+        """
+        used_partition_ids = {p.id for p in get_all_partitions_for_course(block)}
+        return generate_int_id(
+            min_partition_id,
+            max_partition_id,
+            used_partition_ids,
+        )
+
+    @classmethod
+    def fill_teams_user_partitions_ids(cls, block, settings_dict):
+        """
+        Fill the `user_partition_id` in the team settings if it is not set.
+
+        This is used by the Dynamic Team Partition Generator to create the dynamic user partitions
+        based on the team-sets defined in the course.
+        """
+        if not CONTENT_GROUPS_FOR_TEAMS.is_enabled(block.id):
+            return
+
+        proposed_topics = cls.get_team_sets_from_settings(settings_dict)
+        for proposed_topic in proposed_topics:
+            if not proposed_topic.get('user_partition_id'):
+                proposed_topic['user_partition_id'] = cls.get_user_partition_id(
+                    block,
+                    MINIMUM_STATIC_PARTITION_ID,
+                    MYSQL_MAX_INT,
+                )
+
+    @classmethod
+    def get_team_sets_from_settings(cls, settings_dict):
+        """
+        Load team-sets from the course metadata settings.
+        """
+        teams_configuration_model = settings_dict.get('teams_configuration', {})
+        if teams_configuration_model == {}:
+            return {}
+
+        json_value = teams_configuration_model.get('value')
+        if json_value == '':
+            return {}
+
+        proposed_teamsets = json_value.get('team_sets')
+        if proposed_teamsets is None:
+            return {}
+
+        return proposed_teamsets
```

Comment:
We need to make sure this works with `'topics'` as well as `'team_sets'`, for backwards compatibility with older course exports.

I recommend using the `TeamsConfig` class here, which gets you the compatibility for free, plus all the missing-key-fallback logic you have here:
```suggestion
        team_sets = TeamsConfig(settings_dict.get('teams_configuration', {}).get('value')).teamsets
        return [team_set.cleaned_data for team_set in team_sets]
```

Reply:
Mmm, I see. I was avoiding re-implementing how to get teams to avoid adding more complexity, but this looks better. Thank you.

Reply:
No problem!

The `topics` vs `team_sets` thing is an annoying historical thing, and it's kinda my fault that it is that way today--sorry! The important thing to know is that they are synonyms, and we must support them both forever (to avoid breaking OLX), and that the TeamsConfig class is written to handle both keys.

Reply:
So I tried implementing this suggestion, but there's an issue with using TeamsConfig cleaned_data instead of using the settings dictionary directly:

What the team's configuration helpers return is read-only since they return a copy of the objects, so it doesn't work for updating the settings dictionary. So, after returning this list, we still need to look for team_sets or topics (again) in the settings_dict and then update those so the metadata is updated. It's easier to do something similar to the topic fallback, which is repeated several times throughout this file. I understand that repeating this behavior is not ideal, mainly because we're not delegating it to the teams' configuration app, but that will require more refactoring.

I implemented my suggestion in the latest commit: https://github.com/openedx/edx-platform/pull/33788/commits/538d1c97040f910f5d227652038ac16cbadbee62

Reply:
I'm late with this feedback but would it make sense to alter `TeamsConfig` to include a setter method?

Reply:
I thought it was going to be much more work but it was pretty simple: https://github.com/openedx/edx-platform/pull/33788/commits/665a99a828b432f4601ba25cd6a6e4e5f7b762bf

Reply:
Ah, right, TeamsConfig was designed to be immutable. Adding a setter would work, but after re-reading the code, I'd rather take a different approach.

This thread made me re-read this PR's changes to `validate_and_update_from_json`. I knew that `fill_in_teams_user_partitions` was modifying `teams_config` in-place, but that I was perplexed because `teams_config` isn't actually referenced ever again. How were the user_partition_id updates being saved back to the block?? It took me a while to figure out that this works due to a chain of object aliasing:
* `teams_config._data` is aliased to
* `settings_dict["teams_configuration"]` is aliased to
* `filtered_dict["teams_configuration"]` is aliased to
* `key_values["teams_configuration"]`

...and the block updates are applied in the call to `update_from_dict(key_values, block, user)`. Can you see how that chain of effect is hard to follow? I can imagine a future developer accidentally breaking a step in that chain, not realizing that the team partition feature is counting on it.

In cases like this, I find that the code can be made simpler, more robust, and more testable by using the functional paradigm. That is, we pass in only the required data, and we return modified copies rather than updating in-place:
```python
    @classmethod
    def fill_teams_user_partitions_ids(cls, block_id, teams_config):
        """
        Return a new TeamsConfig with each team set's `user_partition_id` filled in.

        This is used by the Dynamic Team Partition Generator to create the dynamic user partitions
        based on the team-sets defined in the course.
        """
        if not CONTENT_GROUPS_FOR_TEAMS.is_enabled(block_id):
            return teams_config
        team_sets = teams_config.teamsets
        for team_set in team_sets:
            if not team_set.user_partition_id:
                team_set.user_partition_id = cls.get_user_partition_id(
                    block,
                    MINIMUM_STATIC_PARTITION_ID,
                    MYSQL_MAX_INT,
                )
        return TeamsConfig({
            **teams_config.cleaned_data,
            "team_sets": team_sets,
        })
```
One note is that this will change the "topics" key to "team_sets", but that is OK, because "team_sets" is the preferred key anyway.

Reply:
Yes, I see how the dict manipulation can be confusing.

I'm implementing a similar change, but I haven't been able to make it work for some reason. I'll let you know when I do.

Reply:
Done! Thank you again for all the suggestions.